jinahub/rankers/SimpleRanker/simpleranker.py

```diff
@@ -1,69 +1,117 @@
 __copyright__ = "Copyright (c) 2020-2021 Jina AI Limited. All rights reserved."
 __license__ = "Apache-2.0"
 
-from itertools import groupby
+from collections import defaultdict
 from typing import Dict, Iterable
 
-from jina import DocumentArray, Executor, requests
+from jina import Document, DocumentArray, Executor, requests
 
 
 class SimpleRanker(Executor):
     """
-    :class:`SimpleRanker` aggregates the score of the matched doc from the
-        matched chunks. For each matched doc, the score is aggregated from all the
-        matched chunks belonging to that doc. The score of the document is the minimum
-        score (min distance) among the chunks. The aggregated matches are sorted by
-        score (ascending).
+    SimpleRanker aggregates the score of matched chunks (matches that are chunks) to the
+    score of the parent document of the matched chunks. The Document's matches are then
+    replaced by matches based on the parent documents of the chunks - they contain an
+    `id` and the aggregated score only.
+
+    This ranker is used to "bubble-up" the scores of matched chunks to the scores
+    of the parent document. As an example, consider an application where we are matching
+    song lyrics to an input query. During indexing we break down all the song lyrics
+    into sentences.
+
+    During querying, we first match lyric sentences to the query, and
+    then use this ranker to produce matching songs (whole songs/lyrics, not just
+    sentences) for the query. Since the matches that this ranker produces contain only
+    document id, we add a final step where an indexer is used to retrieve the song
+    contents base on the id.
     """
 
     def __init__(
         self,
         metric: str = 'cosine',
         ranking: str = 'min',
         traversal_paths: Iterable[str] = ('r',),
+        matches_path: str = 'cm',
```

Comment:
Do we need `travesal_paths` here? It is always from `chunks` to `matches`.

```suggestion
        matches_path: str = 'cm',
```

Reply:
I see you point. We only need to keep either `traversal_paths` or `matches_path`. They are coupled because as long as one of them is fixed, the other is determined.

Reply:
Well, not exactly. Here's the situation:

before my changes, this executor worked like this:
- first, it gathered all the documents that were in `traversal_paths` (call them "query documents")
- for the query documents, if gathered all matches of chunks (`matches_path="cm"`).

I thought that the second step was rather arbitrary and geared towards one of our applications - lyrics search, where the query text is broken into sentences as well, and then we use a searcher to add matches to these sentences (chunks). I saw no reason why you wouldn't want to "merge" matches directly under matches (`matches_path="m"`) of query documents. That's why I added this parameter.

Let me know if this is ok, or if it should be changed. I realize that this makes the executor more complex, but I think that without it it is just too "specific", especially for an executor named "SimpleRanker", which implies it is applicable to a wide range of basic use cases.

Reply:
This is what I imaged about SimpleRanker
```
from jina import Document, DocumentArray
from jina.types.score import NamedScore

K = 3
M = 5

def loop(parent_doc: Document, k):
    chunk = Document(text=f'chunkk, g={k}', granularity=k+1)
    if k == K:
        for i in range(M):
            match = Document(text=f'match {i}')
            match.scores['cosine'] = NamedScore(value=i * 0.1)
            chunk.matches.append(match)
        parent_doc.chunks.append(chunk)
        return parent_doc
    else:
        loop(chunk, k+1)
        parent_doc.chunks.append(chunk)
        return parent_doc


def create_recursive_doc():
    return loop(Document(text='root'), 0)


if __name__ == '__main__':
    docs = DocumentArray([create_recursive_doc()])
    c = docs[0]
    for k in range(K):
        c = c.chunks[0]
    assert len(c.matches) == M

    ranker = SimpleRanker(matches_path='ccccm')
    ranker.rank(docs=docs)
    
    c = docs[0]
    for k in range(K-1):
        c = c.chunks[0]
    assert len(c.matches) == 1
    
```

Reply:
![image](https://user-images.githubusercontent.com/4329072/134445768-ab15b083-1b03-4aa9-ad64-a276aa2c9f1a.png)

Reply:
![image](https://user-images.githubusercontent.com/4329072/134445855-8108ee09-5c32-4d84-8fde-73457eea8034.png)

Reply:
@nan-wang Here is a modified version of your example that accomplishes what you want (according to the diagram you drew). Note the use of `traversal_paths` (which set for which document you want to create aggregated matches) and `matches_path` (which set how to get aggregated matches for that document). Let me know if this is ok


```python
from jina import Document, DocumentArray
from jina.types.score import NamedScore

from simpleranker import SimpleRanker

K = 3
M = 5


def loop(parent_doc: Document, k):
    chunk = Document(text=f'chunkk, g={k}', granularity=k + 1)
    if k == K:
        for i in range(M):
            match = Document(text=f'match {i}', parent_id='1')
            match.scores['cosine'] = i * 0.1
            chunk.matches.append(match)
        parent_doc.chunks.append(chunk)
        return parent_doc
    else:
        loop(chunk, k + 1)
        parent_doc.chunks.append(chunk)
        return parent_doc


def create_recursive_doc():
    return loop(Document(text='root'), 0)


if __name__ == '__main__':
    docs = DocumentArray([create_recursive_doc()])
    c = docs[0]
    for k in range(K + 1):
        c = c.chunks[0]

    assert c.granularity == K + 1  # 4
    assert len(c.matches) == M

    ranker = SimpleRanker(traversal_paths=['ccc'], matches_path='cm')
    ranker.rank(docs=docs)

    c = docs[0]
    for k in range(K):
        c = c.chunks[0]

    assert c.granularity == K  # 3
    assert len(c.matches) == 1
```

Reply:
>         :param matches_path: The path (starting from the "query" document) where the
            matched chunks can be found.
I still think this is an overkill. This makes `SimpleRanker` not simple. 

> I think that without it it is just too "specific", especially for an executor named "SimpleRanker", which implies it is applicable to a wide range of basic use cases.

Being general for basic use cases is fine, but I think we are going too general and makes it over complicated

Reply:
```suggestion
```